[libcxx] Define `_LIBCPP_HAS_C8RTOMB_MBRTOC8` to true if compiling with clang

vhscampos opened the pull request:
Define `_LIBCPP_HAS_C8RTOMB_MBRTOC8` to `1` if compiling with clang.

Some tests involving functionality from `uchar.h`/`cuchar` fail when the platform or the supporting C library does not provide support for the corresponding features. These have been xfailed.

This patch will enable the adoption of newer picolibc versions.

vhscampos committed:
[libcxx] Define _LIBCPP_HAS_C8RTOMB_MBRTOC8 for picolibc

Starting from picolibc 1.8.9, the `char8_t` related functions are
provided.

This patch adds logic to detect the underlying picolibc version and
define the `_LIBCPP_HAS_C8RTOMB_MBRTOC8 macro` accordingly.

vhscampos committed:
Fix clang-format issue

philnik777 reviewed (commented):
I think it would be better to refactor `_LIBCPP_HAS_C8RTOMB_MBRTOC8` so that it's true unless we know it to be false (and also unconditionally make it true with Clang).

vhscampos committed:
Refactor the preprocessor logic

philnik777 commented:
> @philnik777 I don't understand the two suggestions you made. Can you please give more details? Thank you

`_LIBCPP_HAS_C8RTOMB_MBRTOC8` is currently false if we don't know whether a library provides it or not. IMO we should assume it's available unless we know for a fact it's not. With Clang we don't care at all whether it's available, since we have `_LIBCPP_USING_IF_EXISTS`, so we can just always define `_LIBCPP_HAS_C8RTOMB_MBRTOC8` to true in that case.

vhscampos commented:
I've refactored in a way that the macro is true "by default" as you suggested.

However, with regards to always setting it to true if compiling with clang, that causes a libcxx test to fail (link: https://github.com/llvm/llvm-project/blob/372d86dcf1adf90f994be60bb017d92455abb879/libcxx/test/std/strings/c.strings/cuchar.compile.pass.cpp#L26)

The value of macro `TEST_HAS_NO_C8RTOMB_MBRTOC8` depends on `_LIBCPP_HAS_C8RTOMB_MBRTOC8` (link: https://github.com/llvm/llvm-project/blob/372d86dcf1adf90f994be60bb017d92455abb879/libcxx/test/support/test_macros.h#L453), so if we always set it to true for clang, this test fails if the functions aren't defined in the C library.

I couldn't think of a way to modify the test in order to have the macro true for clang.

vhscampos committed:
Handle Apple and MSVC explictly

vhscampos committed:
Handle AIX

vhscampos committed:
Handle FreeBSD and Microsoft C Runtime library

vhscampos committed:
Detect the UCRT library

vhscampos committed:
List of changes:
 - Define macro unconditionally when the compiler is clang.
 - Remove preprocessor logic that isn't needed anymore.
 - Define new libcxx test feature for the AIX platform.
 - Mark tests as xfail in the platforms whose C library doesn't support
   the char8_t functions.

vhscampos committed:
Add xfail for picolibc

vhscampos committed:
List of changes:
 - Move clang's case to the top.
 - Extract char8_t part of the tests into its own files.
 - Remove test that has little use.
 - Undo the creation of a feature just for the AIX platform.

philnik777 reviewed (approved):
LGTM % nit.

vhscampos committed:
Changes:
 - Add android to xfail.
 - Test char8_t unconditionally.

vhscampos committed:
Changes:
 - Add new lit feature to check if glibc is used and if it has char8_t
   support.
 - Use this new lit feature for xfails.

vhscampos committed:
Merge branch 'main' into picolibc-_LIBCPP_HAS_C8RTOMB_MBRTOC8

vhscampos commented:
Since the PR had received approval before my latest changes, I will merge it. If you find any issues with it after the merge, I’m happy to revert it and make any necessary changes.
